akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -230,6 +248,29 @@ private[http2] class Http2ServerDemux(http2Settings: Http2CommonSettings, initia
         }
       })
 
+      /**
+       * Tune this peer to enforce the settings configured from this peer.
+       * @return TODO: it's a Boolean but I don't think it has to be.
+       */
+      private def enforceSettings(settings: immutable.Seq[Setting]): Boolean = {
+        var settingsAppliedOk = true
+
+        settings.foreach {
+          case Setting(Http2Protocol.SettingIdentifier.SETTINGS_MAX_CONCURRENT_STREAMS, value) =>
+          // TODO: Http2Compliance.compliesWithMaxConcurrentStreams should only be enabled once
```

Comment:
Yes, indeed. Especially if you set it to low values on the server, it's not completely unrealistic that the client already sent you a higher number of requests in the meantime which should work.

Reply:
> it's not completely unrealistic that the client already sent you a higher number of requests in the meantime which should work.

Agreed, what I don't like about the spec is that the default value for `SETTINGS_MAX_CONCURRENT_STREAMS` is _infinite_. A malicious client may ignore the server `SETTINGS` frame (or delay the delivery of a `SETTINGS_ACK` enough) and create too many streams. The SETTINGS_ACK timeout should help prevent/hold that kind of attack but enforcing the limit eagerly, as it is currently implemented, sounds good enough for the time being.

Reply:
raised https://github.com/akka/akka-http/issues/3551 so we can improve this in the future.